reframe/frontend/executors/policies.py

```diff
@@ -295,21 +320,11 @@ def _poll_tasks(self):
 
     def _setup_all(self):
         still_waiting = []
-        for task in self._waiting_tasks:
-            if self.deps_failed(task):
-                exc = TaskDependencyError('dependencies failed')
-                task.fail((type(exc), exc, None))
-            elif self.deps_succeeded(task):
-                task.setup(task.testcase.partition,
-                           task.testcase.environ,
-                           sched_flex_alloc_nodes=self.sched_flex_alloc_nodes,
-                           sched_account=self.sched_account,
-                           sched_partition=self.sched_partition,
-                           sched_reservation=self.sched_reservation,
-                           sched_nodelist=self.sched_nodelist,
-                           sched_exclude_nodelist=self.sched_exclude_nodelist,
-                           sched_options=self.sched_options)
-            else:
+        # It needs to loop over a copy of _waiting_tasks because a testcase
+        # could fail on setup and be removed from the list _waiting_tasks.
+        # This would not be not safe and could have unxpected consequences.
+        for task in self._waiting_tasks[:]:
```

Comment:
I don't like this copy. We are also doing another one later on, copying from the `still_waiting`. A more clever way could be to iterate the `waiting_tasks` from left to right, always popping the left item, which you should reinsert at the left if it should still  wait. You stop iterating until you  reach the end of the list. This way you can get rid of the `_remove_from_waiting()` and of the `still_waiting` here.